src/us/kbase/workspace/test/kbase/TypeClientTest.java

```diff
@@ -0,0 +1,462 @@
+import us.kbase.workspace.GetModuleInfoParams;
+import us.kbase.workspace.GrantModuleOwnershipParams;
+import us.kbase.workspace.ListAllTypesParams;
+import us.kbase.workspace.ListModuleVersionsParams;
+import us.kbase.workspace.ListModulesParams;
+import us.kbase.workspace.ModuleInfo;
+import us.kbase.workspace.ModuleVersions;
+import us.kbase.workspace.RegisterTypespecCopyParams;
+import us.kbase.workspace.RegisterTypespecParams;
+import us.kbase.workspace.RemoveModuleOwnershipParams;
+import us.kbase.workspace.TypeInfo;
+import us.kbase.workspace.WorkspaceClient;
+import us.kbase.workspace.kbase.TypeClient;
+import us.kbase.workspace.kbase.TypeDelegationException;
+import us.kbase.workspace.kbase.TypeClient.WorkspaceClientProvider;
+
+public class TypeClientTest {
+
+	// rather than test every method with and without a token, we stagger token usage
+	// same for http vs. https urls
+	
+	// the tests use identity for equals for SDK compiled classes since that's the expected
+	// behavior and doing otherwise would blow up the tests
+	
+	// note the urls are weird as it appears that calling equals on a url actually tries to
+	// access the site. Bad urls seem to be a lot faster than urls that might hit a real
+	// server
+	
+	private static class TestMocks {
+		private final WorkspaceClient wc;
+		private final WorkspaceClientProvider wcp;
+		private final TypeClient tc;
+
+		public TestMocks(
+				final WorkspaceClient wc,
+				final WorkspaceClientProvider wcp,
+				final TypeClient dtsm) {
+			this.wc = wc;
+			this.wcp = wcp;
+			this.tc = dtsm;
+		}
+	}
+	
+	private TestMocks initTestMocks(final URL targetURL) {
+		final WorkspaceClient wc = mock(WorkspaceClient.class);
+		final WorkspaceClientProvider wcp = mock(WorkspaceClientProvider.class);
+		final TypeClient tc = new TypeClient(targetURL, wcp);
+		return new TestMocks(wc, wcp, tc);
+	}
+	
+	@Test
+	public void construct() throws Exception {
+		final TypeClient tc = initTestMocks(new URL("https://cipherisalwaysabadguy.com")).tc;
+		assertThat("incorrect url",
+				tc.getTargetURL(), is(new URL("https://cipherisalwaysabadguy.com")));
+	}
+	
+	@Test
+	public void constructFail() throws Exception {
+		failConstruct(
+				null, mock(WorkspaceClientProvider.class), new NullPointerException("targetURL"));
+		failConstruct(new URL("http://foo.com"), null, new NullPointerException("wsprovider"));
+	}
+	
+	private void failConstruct(
+			final URL url,
+			final WorkspaceClientProvider wcp,
+			final Exception expected) {
+		try {
+			new TypeClient(url, wcp);
+			fail("expected exception");
+		} catch (Exception got) {
+			TestCommon.assertExceptionCorrect(got, expected);
+		}
+	}
+	
+	@Test
+	public void grantModuleOwnership() throws Exception {
+		for (final boolean asAdmin: list(true, false)) {
+			// admin should make no difference
+			final TestMocks m = initTestMocks(new URL("https://whee1111.com"));
+			
+			when(m.wcp.getClient(new URL("https://whee1111.com"))).thenReturn(m.wc);
```

Comment:
YMMV, but I would find it easier and quicker to parse these tests if you set up two variables representing an `https://` URL and an `http://` URL respectively and then used those throughout.

Reply:
I do appreciate that much of the fun in writing the tests would be eliminated by such an URL scheme, though.

Reply:
I tend to prefer to change up variable values among similar tests when it's simple to do so just in case there's some sort of weird behavior that only shows up for a subset of input values. 

It also catches the old
```
def my_func(arg1, arg2):
    return 42 // make tests pass
```

Reply:
And honestly IMO you need to inject as much fun as possible into writing tests to even slightly offset the soul crushing drudgery of writing them